### ISSUE TITLE
No tray icon on Linux (electron@^8)

### ISSUE THREAD

**Issue by vladimiry** [electron/electron#21445]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:** 8.0.0-beta.4
* **Operating System:** Linux (tested on: Manjaro LXDE, Ubuntu 19.04/19.10 Gnome Shell)
* **Last Known Working Electron version:** 7.1.3

### Expected Behavior
Tray icon created.

### Actual Behavior
No tray icon created.

### To Reproduce
https://github.com/vladimiry-playground/electron-quick-start-tray-issue

### Additional Information

The issue has been partially fixed within https://github.com/electron/electron/issues/21154

CC @codebytere

**Comment by codebytere** [comment]
Looks like it's Linux-specific - i was briefly concerned i missed something obvious in testing but it works fine on my macOS when i run your repro. I'll try to look into this more but @zcbenz may also have more answers

**Comment by vladimiry** [comment]
@codebytere thanks for a fast response.

PS The issue is also reproducible on Ubuntu 19.04, the original message got adjuted.

**Comment by codebytere** [comment]
@vladimiry so it turns out the issue isn't quite that it doesn't get created, it's somehow just _very small_ 😆

![image](https://user-images.githubusercontent.com/2036040/70476203-15fe8700-1a8b-11ea-9619-78d149fe981c.png)

See that blue dot? that's it. I'll keep digging since, you know, i don't think anyone is really aiming for teeny tiny icons

**Comment by vladimiry** [comment]
@codebytere it might be so, but on Manjaro I can't even see that small dot. The original image has 128x128 size, https://github.com/vladimiry-playground/electron-quick-start-tray-issue/blob/master/icon.png, referenced to the tray here https://github.com/vladimiry-playground/electron-quick-start-tray-issue/blob/c068c1a20ceebf9bbcc543d51edacbfcbb47f5b1/main.js#L34

>  i don't think anyone is really aiming for teeny tiny icons

Yes please don't discriminate Linux users. The temporary fix could be in shipping tiny icons on all platforms, so there is no discrimination :smile:

**Comment by vladimiry** [comment]
> so it turns out the issue isn't quite that it doesn't get created, it's somehow just very small 

Can confirm such behavior on Ubuntu 19.04, a tiny subtle dot, and it's active - tray menu is working. On Manjaro I can't see the dot nor the functional menu (trying randomly clicking on the taskbar guessing transparent icon is somewhere there).

**Comment by zcbenz** [comment]
The tray icon is implemented with Chromium's `StatusIconLinuxDbus` on Linux, it seems that they have broken something.

**Comment by vladimiry** [comment]
Possibly related chromium's [issue](https://bugs.chromium.org/p/chromium/issues/detail?id=419673) and [commit](https://github.com/chromium/chromium/commit/92171fce24a65a81f90705f974ace3f3dd92c349) (since 79.0.3923.0 tag/version).

**Comment by vladimiry** [comment]
As above-referenced [chromimum issue](https://bugs.chromium.org/p/chromium/issues/detail?id=419673) says `StatusNotifier`/DBus support got added and something got removed. So I looked up things on the internet and got the tray icon working well but only on Manjaro XFCE so far (have not tried anything KDE based yet though, might be working well there):
- Manjaro XFCE (possibly any distro that uses XFCE, like Xubuntu): `xfce4-statusnotifier-plugin` needs to be installed and then respective widget added to the panel. So `xfce4-panel`/built-in systray plugin named "notification area" is not sufficient anymore. Plugins work well if both activated, no conflicts noticed.
- Ubuntu 19.04 Gnome Shell: [KStatusNotifierItem/AppIndicator Support](https://extensions.gnome.org/extension/615/appindicator-support/) extension is supposed to handle the need as it seems to be supporting "status notifier/dbus" thing. But it doesn't work well, there is still that subtle tiny dot instead of a full tray icon. So further digging is needed.

The side benefit I noticed is that right mouse click is being properly handled again after it's being broken starting from the @electron v3, see #14941. So now right click and left click are treated as a different action. Besides `xfce4-statusnotifier-plugin` is capable enough to optionally cancel this/proper behavior, so if someone needs it both right and left clicks still could trigger a single action - menu opening (it's good that the respective checkbox is turned off by default so we have right and left clicks working individually).

Since for now, there is no fully reliable way to guarantee that the tray icon is properly shown on Linux using electron v8+ I'm considering making "startMinimizedToTray" option disabled by default in [ElectronMail](https://github.com/vladimiry/ElectronMail). I will then add a note `make sure you have got the tray icon displayed before enabling this option as you will be using tray icon to show/hide the app window` near to the respective checkbox in the UI. That would not be the best move in terms of UX but it's better than if a user starts the app and doesn't see it started as there is no tray icon and so no way to get the app window displayed, basically no visible signs of the app got started.

**Comment by vladimiry** [comment]
Since I finally got the tray icon displayed on Linux in one case using @electon version 8 I played around for a while and noticed a few more issues which I'm going to place separately later if I will be able to narrow down the scope and create minimal reproducible projects for them. The following issues (based on preliminary observations):
- It looks like if the async/delayed action is being triggered via the tray menu the app crashes/segfaults. Possibly any kind of actions, not only the async ones, not completely clear yet. If I trigger the same action not via the tray menu then there is no crush.
- The tray menu items don't get updated dynamically as it worked in @electon version lower than 8, at least the `visible` property (I created a few hidden by default menu items and then switch the visibility dynamically as needed).

These issues are off the topic here but wanted to highlight that in general, it's more than just tray icon not being displayed on Linux but the tray menu behaviour is also changed.

**Comment by vladimiry** [comment]
> noticed a few more issues which I'm going to place separately later

#21499

**Comment by vladimiry** [comment]
Kubuntu 19.10 (fresh install): icon well displayed.

**Comment by vladimiry** [comment]
I just resized the icon from 128x128 down to 32x32 and it got displayed in full size on Ubuntu 19.10 Gnome Shell (fresh install + all updates). So there are ways to make the tray icon displayed on all the systems I tested the case (Ubuntu Gnome Shell, Kubuntu, Manjaro XFCE) and so I don't expect Electron will need to fix something about tray icon displaying but would prefer the issue to remain open for a while as many developers will face the same issue switching to Electron version 8.

**Comment by vladimiry** [comment]
Yes this issue is not a blocker, we figured out how to make the icon to be visible on all the popular kinds of Linux, so I'm closing it. There clearly was a major change on how the tray icon on Linux gets displayed, but this change seems to be coming from Chromium.

PS But #21499 is a real blocker for me even though it's not labeled as such.

**Comment by ckerr** [comment]
This is still an issue and should probably be kept open until resolved

**Comment by vladimiry** [comment]
@ckerr did I get it right, Electron will be downscaling the tray image automatically on Linux if needed (as I guess 32x32 size I experimentally detected in theory might differ on different systems)?

**Comment by vladimiry** [comment]
Can confirm in recent beta release even if bigger than 128x128 icon is used it still being well displayed on Ubuntu. So yes Electron now handles the downscaling internally if needed.

**Comment by squalou** [comment]
Hi,

just to mention that it does not work with Cinnamon (on latest Mint for instance)

Didn't find any plugin thaht enables it, if someone knows of one .... thank you.

**Comment by vladimiry** [comment]
> just to mention that it does not work with Cinnamon (on latest Mint for instance)

https://github.com/polybar/polybar panel usually used with i3 tiling window manager also doesn't seems to support StatusNotifier tray icons.

**Comment by vladimiry** [comment]
Linux Mint/Cinnamon doesn't support StatusNotifier tray icons standard. So it's an upstream issue you might want to place in Mint/Cinnamon issues tracker.

**Comment by dotenorio** [comment]
See my solution:

https://github.com/zulip/zulip-desktop/issues/910#issuecomment-608056150

**Comment by abhigyank** [comment]
Latest beta release still doesn't fix this for me on Ubuntu 16.04 and GNOME Shell 3.18.5.
Is there a fix for that, if not can this issue be reopened?

**Comment by vladimiry** [comment]
To whom it may concern. It turned out Electron is getting rid of DBus related tray icon implementation rolled out with @chromium update in Electron v8 in favor of previously used / `GTK/appindicator` approach, see electron/electron#23674.

**Comment by hovancik** [comment]
I am actually sorry to hear the `StatusIconLinuxDbus` is getting ripped of. For me, it solved many tray related issues and my app was finally working well on Linux ;/

**Comment by akontsevich** [comment]
> I am actually sorry to hear the `StatusIconLinuxDbus` is getting ripped of. For me, it solved many tray related issues and my app was finally working well on Linux ;/

Yes, solved for me as well. Very strange and stupid revert.

**Comment by bkmo** [comment]
Sorry to hear that the Dbus tray icon implementation is gone. It was great when it was implemented and the tray function finally worked as designed.

**Comment by hovancik** [comment]
Yep, tray tooltip was finally working for for me (XFCE, KDE) - which were broken before. Also, tray icon was finally looking good.

I am not sure why it was reverted, I have only had things **started to work**, and all it took was to install some extra packages mentioned in this thread to make new implementation to work. Seems pretty unreasonable to go back to old implementation when new was working better and all it took was some extra package to install...

**Comment by akontsevich** [comment]
I think the problem is they just do not use features they are developing and therefore do not understand functionality state and reverting 100% working things according to strange or outdated bug reports.

**Comment by squalou** [comment]
IMHO, there are *plenty* DE out there, each implementing one or another one so-called 'standard' to deal with the systray paradigm. This is a mess, yet this is real life.

Why electron can't keep both implementation instead of these violent switches bothers me.

(I mean : it's electron that pretends to address portability to all OS's and OS flavours, it's not an easy task, but that's what the project is about).

**Comment by maruthi-adithya** [comment]
@vladimiry @hovancik I tried what you had mentioned in this [comment](https://github.com/electron/electron/issues/21445#issuecomment-565710027). I have XFCE based XUbuntu OS. When I install `xfce4-statusnotifier-plugin`, it says `xfce4-statusnotifier-plugin is already the newest version (0.2.1-0ubuntu1).` When I run my electron app, with a simple tray initialization, I am not seeing any tray. Can you please tell me what I am missing in this?

**Comment by calidion** [comment]
Tray is still not working with electron 9.1.1 on Ubuntu 18.04

**Comment by mishuvs** [comment]
I wrote the following code for my main process:

```
const iconPath = path.join(__dirname, 'iconTemplate.png');
let tray = null;

app.whenReady().then(() => {
    tray = new Tray(iconPath);
    tray.setToolTip('Alarm');
})
```
I didn't add the context menu because I had no use. Though I tried adding just in case it was required but it still didn't work. I am not using any rendering process either.

I am using `auto-launch` package as well.

Here's how my tray looks:

![Screenshot from 2020-08-25 14-47-20](https://user-images.githubusercontent.com/35009811/91156571-ec9d8e00-e6e1-11ea-9ce9-768862de18ac.png)

I tried the tray demo from [electron-demo-api](https://github.com/electron/electron-api-demos) and the [tray icon appears](https://github.com/electron/electron-api-demos/issues/460) for the demo, so I am pretty sure it's not an OS problem. Any ideas on how to fix it?
